File: Week_02/589.n叉树的前序遍历.py
#
# @lc app=leetcode.cn id=589 lang=python3
#
# [589] N叉树的前序遍历
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""


class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        stack, res = [root], []
        while stack:
            root = stack.pop()
            res.append(root.val)
            for child in root.children[::-1]:
                stack.append(child)
        return res
# @lc code=end

File: Week_05/589.n叉树的前序遍历.cpp
/*
 * @lc app=leetcode.cn id=589 lang=cpp
 *
 * [589] N叉树的前序遍历
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution
{
public:
    vector<int> preorder(Node *root)
    {
        vector<int> res;
        if (root == nullptr)
            return res;
        stack<Node *> st;
        st.push(root);
        while (!st.empty())
        {
            Node *node = st.top();
            res.push_back(node->val);
            st.pop();
            size_t n = node->children.size();
            for (int i = n - 1; i >= 0; i--)
            {
                if (node->children[i] != nullptr)
                {
                    st.push(node->children[i]);
                }
            }
        }
        return res;
    }
};
// @lc code=end
